Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp

```diff
@@ -356,48 +430,89 @@ RetainPtr<CTFontRef> UnrealizedCoreTextFont::realize() const
-    // opportunistically assume the font at hand is one of those and set the values accordingly. Now,
-    // after we've created the font, we can check to see whether or not our guess was correct. If it
-    // wasn't, we have to recreate the font with the normalized values.
-    //
-    // Once rdar://problem/105483251 is solved, we can delete this section.
-    if (m_applyTraitsVariations == ApplyTraitsVariations::Yes && FontInterrogation(font.get()).variationType == FontInterrogation::VariationType::TrueTypeGX) {
-        auto variationValues = defaultVariationValues(font.get(), ShouldLocalizeAxisNames::No);
-        if (variationValues.contains({ { 'w', 'g', 'h', 't' } })
-            || variationValues.contains({ { 'w', 'd', 't', 'h' } })
-            || variationValues.contains({ { 'i', 't', 'a', 'l' } })
-            || variationValues.contains({ { 's', 'l', 'n', 't' } })) {
-            VariationsMap variationsToBeApplied;
-
-            auto weight = denormalizeGXWeight(m_weight);
-            auto width = denormalizeVariationWidth(m_width);
-            auto slope = denormalizeSlope(m_slope);
-
-            variationsToBeApplied.set({ { 'w', 'g', 'h', 't' } }, weight);
-            variationsToBeApplied.set({ { 'w', 'd', 't', 'h' } }, width);
+    if (auto rebuildReason = this->rebuildReason(font.get()); rebuildReason.hasEffect()) {
+        VariationsMap variationsToBeApplied;
+        std::optional<VariationDefaultsMap> variationDefaults = WTFMove(rebuildReason.variationDefaults);
+#if USE(CORE_TEXT_VARIATIONS_CLAMPING_WORKAROUND)
+        // Even if everything was in-bounds, we still have to clamp,
+        // because normalization might put us out-of-bounds.
+        if (!variationDefaults)
+            variationDefaults = defaultVariationValues(font.get(), ShouldLocalizeAxisNames::No);
+#endif
+
+        auto maybeClampToRange = [&variationDefaults](FontTag fontTag, float value) -> std::optional<float> {
+#if USE(CORE_TEXT_VARIATIONS_CLAMPING_WORKAROUND)
+            ASSERT(variationDefaults);
+            auto iterator = variationDefaults->find(fontTag);
+            if (iterator == variationDefaults->end()) {
+                // The font doesn't support this variation axis.
+                return { };
+            }
+            return iterator->value.clamp(value);
+#else
+            return value;
+#endif
+        };
+
+        auto setVariation = [&variationsToBeApplied, &maybeClampToRange](FontTag fontTag, float value) {
+            if (auto clamped = maybeClampToRange(fontTag, value))
+                variationsToBeApplied.set(fontTag, value);
+        };
+
+        auto weight = m_weight;
+        auto width = m_width;
+        auto slope = m_slope;
+
+        if (rebuildReason.gxVariations) {
+            weight = denormalizeGXWeight(m_weight);
+            width = denormalizeVariationWidth(m_width);
+            slope = denormalizeSlope(m_slope);
+        }
+
+        {
```

Comment:
Nit: this block doesn't seem to be needed.

Reply:
I think I'd rather keep it because it gives the reader a visual grouping of the various pieces of this function.